### PR TITLE
Provider refactoring (and documentation update)

### PR DESCRIPTION
**Pull request by rcherrueau**
Provider modifications are:
- Removes old before/after pre-install
- Puts out the loading of configuration
- Add automatic checks for provider keys marked as required
- Abstracts the interpretation of a topology

### CODE REVIEW

**Comment by rcherrueau** [comment]
Tested for G5k and vagrant. @msimonin could you take a look for openstack/chameleon?

File: enos/provider/chameleonkvm.py [hunk 1]
```diff
@@ -1,19 +1,20 @@
...
-    def init(self, config, calldir, force_deploy=False):
-        conf = load_config(config, default_provider_config=PROVIDER_CONFIG)
+    def init(self, conf, calldir, force_deploy=False):
         return super(Chameleonkvm, self).init(conf, calldir, force_deploy)
 
     def destroy(self, calldir, env):
         super(Chameleonkvm, self).destroy(calldir, env)
+
+    def default_config(self):
+        return super.defautl_config().update({
```
Comment (msimonin) [inline enos/provider/chameleonkvm.py]: typo : `default_config()`

Comment (rcherrueau) [inline enos/provider/chameleonkvm.py]: Done

File: enos/provider/chameleonkvm.py [hunk 2]
```diff
@@ -1,19 +1,20 @@
...
-    def init(self, config, calldir, force_deploy=False):
-        conf = load_config(config, default_provider_config=PROVIDER_CONFIG)
+    def init(self, conf, calldir, force_deploy=False):
         return super(Chameleonkvm, self).init(conf, calldir, force_deploy)
 
     def destroy(self, calldir, env):
         super(Chameleonkvm, self).destroy(calldir, env)
+
+    def default_config(self):
+        return super.default_config().update({
```
Comment (msimonin) [inline enos/provider/chameleonkvm.py]: Actually this should be  : 
```
        default_config = super(Chameleonkvm, self).default_config()
        default_config.update({
            'type': 'chameleonkvm',
            'image_name': 'CC-Ubuntu16.04',
            'user': 'cc',
            'dns_nameservers': ['129.114.97.1',
                                '129.114.97.2',
                                '129.116.84.203'],
            'network_interface': 'ens3'
        })
        return default_config
```

Comment (rcherrueau) [inline enos/provider/chameleonkvm.py]: Arf my double bad. First, I don't know how to make inheritance in python. Then, I forgot that the `update` returns `None`. I am not used to these mutable interfaces :weary:

Comment (rcherrueau) [inline enos/provider/chameleonkvm.py]: Done

File: docs/contribute/new-provider.rst [hunk 3]
```diff
@@ -28,9 +28,11 @@ with:
 
    # Use rsc to book resources ...
 
-At the end of the ``init``, the provider should return a list of hosts
-that Enos can SSH on, together with a pool of available IP for
-OpenStack Network.
+At the end of the ``init``, the provider should return a list of
+`host.py_` that Enos can SSH on, together with a pool of available IP
```
Comment (msimonin) [inline docs/contribute/new-provider.rst]: `
`host.py`_
`

File: docs/provider/index.rst [hunk 4]
```diff
@@ -3,12 +3,18 @@
...
 =========
 
-Enos is built around the notion of provider. It lets you use Enos on various
-infrastructure.
+Enos offers to test different OpenStack deployments over some
+resources. In the context of Enos, a resource is anything Enos can SSH
+on and start a Docker daemon. Thus, a resource could be a bare-metal
+machine, a virtual machine, or a container resource depending on the
+testbed used for conduction the experiments. To get these resources,
+Enos relies on a notion of *provider* and already implements
```
Comment (msimonin) [inline docs/provider/index.rst]: implements *the* followings ?

File: docs/contribute/new-provider.rst [hunk 5]
```diff
@@ -8,8 +8,8 @@ The actual implementation gives two providers: :ref:`grid5000` and
 then implementing a new one is easy as 500 lines of Python code.
```
Comment (msimonin) [inline docs/contribute/new-provider.rst]: Maybe fixe the above sentence : 
`
The actual implementation gives you several providers, ...
`

Comment (rcherrueau) [inline docs/contribute/new-provider.rst]: Actually, I fixed it in the next #147. But during my process of splitting #147 in two, I cannot manage to edit the diff file to remove the reference to the static provider. So I decided to leave this paragraph has it is, with the knowledge that a fix is coming in #147.

Comment (msimonin) [inline docs/contribute/new-provider.rst]: cool

File: docs/provider/index.rst (hunk 4 above)
Comment (rcherrueau) [inline docs/provider/index.rst]: Done

File: docs/contribute/new-provider.rst (hunk 3 above)
Comment (rcherrueau) [inline docs/contribute/new-provider.rst]: Done

**Pull request merged**